### DISCUSSION TITLE
CP Sat is not optimizing my cost function

### DISCUSSION THREAD

**Discussion by latkm** [google/or-tools#2491] [CP-SAT questions] [answered]
Hi,

I'm setting up a constrained optimization problem, with boolean vars, in CP Sat.
I notice that the returned solution satisfy all the constraints but returned "optimal" value f(x*) for objective function (ObjectiveValue), at the found "optimal" argument (x*) is not the real value for the objective function. E.g. f(x*) returned is 0, while the real value for f(x*) should be 23434. When I evaluate f(x*) by using the same function defined for computing the objective function, I do get the correct value. But when it is called by CP Sat through the optimization process, it seems to be computing wrongly. 

I post the definition of the objective function, which calls other previously defined functions.

```python
def cost_function(z):

    value = 0
    tc = paths_of_z(z)
    time_requirements = read_times(tc, time_tables)
    cn = cost_of_paths(tc, costs)
    for m in np.arange(0, len(cn), 1):
        value = value + sum((np.array(cn[m]) - np.array(time_requirements[m]))**2)
    return value
model.Minimize(cost_function(z))
```

In order to debug it, I'm trying to print the intermediate solutions, i.e. solutions attained during the sequence of iterations, with:
(I'm not sure if intermediate solutions, can be printed - please correct if I'm wrong)
 
```python
class VarArrayAndObjectiveSolutionPrinter(cp_model.CpSolverSolutionCallback):
    """Print intermediate solutions."""

    def __init__(self, variables):
        cp_model.CpSolverSolutionCallback.__init__(self)
        self.__variables = variables
        self.__solution_count = 0

    def on_solution_callback(self):
        print('Solution %i' % self.__solution_count)
        print('  objective value = %i' % self.ObjectiveValue())
        for v in self.__variables:
            print('  %s = %i' % (v, self.Value(v)), end=' ')
        print()
        self.__solution_count += 1

    def solution_count(self):
        return self.__solution_count
        # [END print_solution]`
```
But I get only one solution printed, which seems to be the first guess that is feasible, and for which the value f(x*) is wrongly computed.

Any help will be greatly appreciated. 
Thank you very much,

**Comment by lperron** [comment, marked as answer]
You cannot pass an arbitrary function to the minimize code. 

In particular **2 is not supported.

**Reply by latkm** [reply]
Thanks for your input. 
Is there documentation about what functions can be passed to the minimize code, and what math operations are supported?
Where I can read about that?

Thanks again!

**Reply by Mizux** [reply]
a good start, take a look of all operators supported in the LinearExpr class
-> https://github.com/google/or-tools/blob/fa84bc05e72641dddfbb98164d81b8bc9bef6ea5/ortools/sat/python/cp_model.py#L203-L313

**Reply by Mizux** [reply]
Maybe we should redefine `__pow__`
don't know where to find the exhaustive list of python operator
EDIT: maybe [here](https://docs.python.org/3/library/operator.html#operator.__pow__)

**Reply by ghost** [reply]
Here: https://docs.python.org/3/reference/datamodel.html#emulating-numeric-types
I also think that you should take a look at  `__bool__` so that people stop doing:
```python
if var == value:
    ...
```
the downside it would also trigger the warning:
```python
var = None
var = model.New...

if var:
    ...
```

**Reply by latkm** [reply]
Thank you very much @Mizux! As per the x**2 operator, could it be replaced by (x)*(x)? 
Are the functions to be optimized restricted to be linear expressions of the variables? Or we are allowed to define more involved objective functions?

**Reply by lperron** [reply]
The objective function needs to be linear. But some variables in the linear equations can be the result of a product of two affine expressions.

**Comment by lperron** [comment]
At least it should throw an error.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by latkm** [reply]
The fact is that it doesn't return any error. The computation goes smoothly as if the objective function were properly defined. I guess that it confuses me.

**Comment by lperron** [comment]
I have pushed a modification to cp_model.py

```python
if var == 3:

if var >= 3:

model.Add(x == min(y, z))  # 3 IntVar.
```

All will print a warning.

I am hesitant about `IntVar.__bool__()`
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le mar. 13 avr. 2021 à 17:24, Xiang Chen ***@***.***> a
écrit :

**Comment by lperron** [comment]
No, the objective is linear.
Now, you can use AddMultiplicationEquality, AddMinEquality, or
AddMaxEquality() to link variables to more complex expressions.


Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by latkm** [reply]
Hello,

Thank you all for the provided help. I'm still struggling to understand if the CP-Sat framework its appropriate for my optimization problem. The problem is basically a routing problem, for which we want to formulate certain constraints that, at the best of our understanding, cannot be formulated with the commands provided in the OR Tools for VRP.  We thought that this framework could be more generic allowing us to formulate more generic constraints in terms of boolean variables that represent whether or not a given arc is part of the optimal graph obtained for the routing problem. So far, we are good with the constraint formulations.
However, with respect to the objective function we need to consider cumulative times of the trips, for instance the cumulative time until node 3, and the cumulative time until node 4, and so on. I wrote python functions to compute these quantities based on the independent variables (optimization variables). But, as per comments above in this post, I understand that I cannot optimize any arbitrary function of the independent variables. 

Are there any specific methods within the CP Sat framework that would allow me to express cumulative quantities of the independent variables, in such a way that these expressions could be part of the objective function?

I'll highly appreciate any input in this respect.  

L

**Reply by Mizux** [reply]
Did you try `model.Minimize(cp_model.LinearExpr.Sum(expressions))` ?

**Comment by lperron** [comment]
you need a circuit constraint. And a cumul var per node.

arc between node i and node j is true iff j is the direct successor of i.
Then you can use a reified linear expression to do the accumulation


model.Add(cumul_var[j] == cumul_var[i] +
transit[i][j]).OnlyEnforceIf(arc_i_j)

note that the == can be a >= if you can have some slack.

With this basic idea, you should be able to express your problem.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by latkm** [reply]
But how could you use the AddCircuit method to generate a constraint if a priori you don't know which potential arcs of the graph will be part of the optimal solution? We know what arcs are part of the graph, once we know the optimal solution. 
With that in mind, is there still a way to do it using AddCircuit?

**Comment by lperron** [comment]
You need to add one reified equation per arc.
